remove openssl dependency

drahnr opened the pull request:
Continuation of #1742 

OpenSSL is a hinderance of both building and distribution, and has been superseeded by rustls and boringssl as safer and alternatives with sufficient subsets for sccache. We hance should not rely on a dependency when there are easier in alternatives.

Particularly difficult use cases when linking against OpenSSL:

* version mismatches
* static linkage
* cross compilation
* certificate store configuration

NobodyXu committed:
dep: Upgrade trust-dns-resolver v0.22 to hickory-resolver v0.24

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

NobodyXu committed:
Fix line ending of dynamic key generated

Use CRLF on windows and `\n` on Linux.

Also fix formatting of `Cargo.toml`

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

NobodyXu committed:
Fix typo: `Default()` => `default()`

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

drahnr committed:
fix: toml formating

sylvestre commented:
In the commit, it would be nice to explicit why we want to remove it :)
Thanks for restarting this

drahnr committed:
add certificate equiv test, move to sha256, newer openssl distributions do not contain sha1 support by default

drahnr committed:
clippy

drahnr commented:
Done

drahnr committed:
fmt again

drahnr committed:
remvoe flipper RNG + guard test module to dist-{client,server}

drahnr committed:
don't add mask to ip address authority

drahnr commented:
The issue is somewhere in the delta:

```
# openssl
[
Extension { extn_id: ObjectIdentifierAsn1(ObjectIdentifier { root: JointIsoItuT, first_node: 5, child_nodes: [29, 17] }), critical: Implicit(false), extn_value: SubjectAltName(OctetStringAsn1Container(Asn1SequenceOf([IpAddress(OctetStringAsn1([127, 0, 0, 1]))]))) },
Extension { extn_id: ObjectIdentifierAsn1(ObjectIdentifier { root: JointIsoItuT, first_node: 5, child_nodes: [29, 37] }), critical: Implicit(false), extn_value: ExtendedKeyUsage(OctetStringAsn1Container(ExtendedKeyUsage(Asn1SequenceOf([ObjectIdentifierAsn1(ObjectIdentifier { root: Iso, first_node: 3, child_nodes: [6, 1, 5, 5, 7, 3, 1] })])))) }
]
 
# picky
[
Extension { extn_id: ObjectIdentifierAsn1(ObjectIdentifier { root: JointIsoItuT, first_node: 5, child_nodes: [29, 17] }), critical: Implicit(true), extn_value: SubjectAltName(OctetStringAsn1Container(Asn1SequenceOf([IpAddress(OctetStringAsn1([127, 0, 0, 1]))]))) }
Extension { extn_id: ObjectIdentifierAsn1(ObjectIdentifier { root: JointIsoItuT, first_node: 5, child_nodes: [29, 37] }), critical: Implicit(true), extn_value: ExtendedKeyUsage(OctetStringAsn1Container(ExtendedKeyUsage(Asn1SequenceOf([ObjectIdentifierAsn1(ObjectIdentifier { root: Iso, first_node: 3, child_nodes: [6, 1, 5, 5, 7, 3, 1] })])))) }
Extension { extn_id: ObjectIdentifierAsn1(ObjectIdentifier { root: JointIsoItuT, first_node: 5, child_nodes: [29, 19] }), critical: Implicit(false), extn_value: BasicConstraints(OctetStringAsn1Container(BasicConstraints { ca: Implicit(Some(false)), path_len_constraint: Implicit(None) })) }
Extension { extn_id: ObjectIdentifierAsn1(ObjectIdentifier { root: JointIsoItuT, first_node: 5, child_nodes: [29, 35] }), critical: Implicit(false), extn_value: AuthorityKeyIdentifier(OctetStringAsn1Container(AuthorityKeyIdentifier { key_identifier: Some(ContextTag0(OctetStringAsn1([53, 4, 192, 142, 244, 179, 95, 190, 50, 190, 70, 215, 188, 117, 9, 37, 50, 147, 131, 167, 10, 120, 7, 227, 65, 73, 228, 209, 66, 145, 160, 107]))), authority_cert_issuer: None, authority_cert_serial_number: None })) }
Extension { extn_id: ObjectIdentifierAsn1(ObjectIdentifier { root: JointIsoItuT, first_node: 5, child_nodes: [29, 14] }), critical: Implicit(false), extn_value: SubjectKeyIdentifier(OctetStringAsn1Container(OctetStringAsn1([53, 4, 192, 142, 244, 179, 95, 190, 50, 190, 70, 215, 188, 117, 9, 37, 50, 147, 131, 167, 10, 120, 7, 227, 65, 73, 228, 209, 66, 145, 160, 107]))) }
]
```

The basic fields differ in `Implicit(true)` and `Implicit(false)`, but picky adds 3 extra attributes in `CertificateBuilder::build` extra extended fields and I have yet to read the spec on it

drahnr committed:
use vendored openssl for tests

drahnr committed:
toml fmt

AJIOB commented on Cargo.toml:
```diff
@@ -128,11 +138,14 @@ cc = "1.0"
 chrono = "0.4.33"
 filetime = "0.2"
 itertools = "0.12"
+openssl = { version = "=0.10.64", features = ["vendored"] }
```
Do we really need to embed the OpenSSL too? Why not just fully rely on rustls?

drahnr committed:
clippy was wrong

drahnr commented:
The remaining issues originate from https://github.com/briansmith/ring/issues/1167#issuecomment-1745560468 and will be mitigated once `hickory-dns` gets a `0.25.0` release https://github.com/hickory-dns/hickory-dns/issues/2206

drahnr committed:
dep: force `hickory-dns >= 0.24.4` or higher

Uses `ring >= 0.7.0` which contains a fix for win on aarch64

drahnr committed:
ignore openssl cert congruence test

It will never work with `picky`, the `openssl` generated certs were invalid
in the first place.

drahnr committed:
fmt

drahnr commented:
`hickory-dns` `0.24.4` got released

drahnr committed:
lock

drahnr committed:
remove openssl integrity test, we can't ensure it

Simplifies CI

AJIOB commented on src/compiler/c.rs:
```diff
@@ -286,7 +286,7 @@ where
             .map(|f| {
                 f.flatten()
-                    .filter(|f| f.path().extension().map_or(false, |ext| ext == "bc"))
+                    .filter(|f| f.path().extension().is_some_and(|ext| ext == "bc"))
```
What about migrating OpenSSL-unrelated changes to the dedicated PR?
I don't think that changes breaks CI, but this split will minify change set, that is related to failure.

drahnr replied:
Sure

drahnr commented:
I cannot reproduce `test_dist_restartedserver` issues CI is encountering. I've seen over PRs where re-trying the CI ultimately succeeded.

drahnr committed:
ensure dist tests are run serially and don't kill each others local daemon

drahnr commented:
I can occassionally repro the issue locally

drahnr committed:
feat: remove openssl dependency

 - Bump rouille from 3.5 => 3.6.2
   rouille v3.6.2 fixed a bug: `rouille::Server::new_ssl` is now exposed
   when only `rustls` is enabled.
 - Disable default features of `reqwest`
   which pulls in openssl
 - Remove `openssl` pulled in `[dev-dependencies]`
 - Bump reqwest from 0.11.17 => 0.11.18

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

drahnr committed:
lock

drahnr committed:
lock again

drahnr committed:
fix/ci: grcov should be installed with locked to avoid unstable features
